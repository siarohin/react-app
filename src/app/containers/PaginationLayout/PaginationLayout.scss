.app-pagination {
  *[class*="MuiButtonBase-root"],
  *[class*="MuiPaginationItem-root"] {
    color: var(--app-theme-grey-light-25);
    font-family: var(--app-font-family);
    font-size: 1.4rem;
    height: 4rem;
    width: 4rem;
    border-radius: 50%;

    &:hover:not([class*="MuiPaginationItem-ellipsis"]) {
      border: 0.1rem solid var(--app-theme-data-viz-red);
      color: var(--app-theme-data-viz-red);
    }

    &[class*="MuiPaginationItem-ellipsis"] {
      font-size: 1.8rem;
      height: auto;
    }

    &[class*="Mui-selected"],
    &[class*="Mui-selected"]:hover {
      color: var(--app-theme-grey-light-25);
      background: var(--app-theme-data-viz-red);
    }
  }
}
