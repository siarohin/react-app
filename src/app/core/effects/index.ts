import { combineEpics } from "redux-observable";

import { movieEffects } from "./movies.effect";
import { userPreferencesEffects } from "./user-preferences.effect";
import { toastEffects } from "./toast.effect";

export * from "./common.util";

export const effects = combineEpics(movieEffects, userPreferencesEffects, toastEffects);
