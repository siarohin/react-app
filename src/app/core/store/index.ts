export * from "./reducers";
export { SharedModels } from "./shared";
export { MoviesActions, MoviesModels } from "./movies";
export { UserPreferencesActions, UserPreferencesModels } from "./user-preferences";
export { ToastActions } from "./toast";
