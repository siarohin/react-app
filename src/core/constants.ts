/**
 * App name
 */
export const APP_NAME: string = "React movie";

/**
 * Dialog actions
 */
export enum DialogAction {
  CREATE = "Create",
  UPDATE = "Update",
  DELETE = "Delete"
}

/**
 * Dialog titles
 */
export enum DialogTitle {
  CREATE = "Add movie",
  UPDATE = "Edit movie",
  DELETE = "Delete movie"
}

/**
 * Selectable filter options
 */
export enum FilterOptions {
  rating = "Rating",
  releaseDate = "Release date"
}

/**
 * Queries that are used in the filter request
 */
export enum FilterOptionsQuery {
  rating = "vote_average",
  releaseDate = "release_date"
}

/**
 * Filter genres
 */
export const FILTER_GENRES: Array<string> = ["Comedy", "Crime", "Documentary", "Drama", "Horror", "Romance"];

/**
 * Default filter value
 */
export const DEFAULT_GENRE: string = "All";

/**
 * All the backend API endpoint URIs are listed here.
 */
export enum Endpoints {
  // REST APIs:
  AppHost = "http://localhost:4000"
}

/**
 * Movies per page
 */
export const MOVIES_PER_PAGE: number = 30;

/**
 * Sorting order
 */
export const SORTING_ORDER: "asc" | "desc" = "desc";

/**
 * Search option
 */
export const SEARCH_BY: "title" | "genres" = "title";

/**
 * Upsert movie messages enum
 */
export enum UpsertMovieMsg {
  Success = "The movie has been successfully saved.",
  Fail = "The movie could not be saved."
}

/**
 * Delete movie messages enum
 */
export enum DeleteMovieMsg {
  Fail = "The movie could not be deleted.",
  Success = "The movie has been successfully deleted."
}
