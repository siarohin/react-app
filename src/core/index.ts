export * from "./constants";
export * from "./services";
export * from "./effects";
export * from "./store";
export * from "./store-configuration";
