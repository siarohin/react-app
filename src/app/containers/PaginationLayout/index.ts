import PaginationLayout from "./PaginationLayout";

export { PaginationLayout };
