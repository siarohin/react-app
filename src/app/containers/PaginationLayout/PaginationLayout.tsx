import React, { ChangeEvent, ComponentType, ReactElement, useMemo } from "react";
import { connect } from "react-redux";

import "./PaginationLayout.scss";
import { Pagination } from "../../shared";
import { MoviesModels, MOVIES_PER_PAGE, State, SharedModels, UserPreferencesActions } from "../../core";

const PaginationLayout = (
  props: Partial<MoviesModels.IMoviesState> & SharedModels.IDispatchAction
): ReactElement<any> | null => {
  const { totalAmount, offset, limit, dispatch } = props;

  const count = useMemo((): number => Math.ceil(totalAmount! / limit!), [totalAmount, limit]);
  const page = useMemo((): number => offset! / limit! + 1, [offset, limit]);

  const handleOnChange = (event: ChangeEvent<unknown>) => {
    const newOffset: number = Number((event.target as HTMLElement).innerText) * MOVIES_PER_PAGE - MOVIES_PER_PAGE;
    const shouldUpdateOffset: boolean = newOffset !== offset;

    if (shouldUpdateOffset) {
      dispatch(UserPreferencesActions.updateOffset({ selected: newOffset.toString() }));
    }
  };

  const isPaginationVisible: boolean = count > 1;

  if (!isPaginationVisible) {
    return null;
  }

  return (
    <Pagination
      className="app-pagination"
      onChange={handleOnChange}
      page={page}
      count={count}
      hidePrevButton
      hideNextButton
    />
  );
};

const mapStateToProps = (state: State): Partial<MoviesModels.IMoviesState> => {
  return {
    totalAmount: state.movieList.totalAmount,
    limit: state.movieList.limit,
    offset: state.movieList.offset
  };
};

export default connect(mapStateToProps)(PaginationLayout as ComponentType<any>);
