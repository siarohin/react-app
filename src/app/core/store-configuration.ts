import { applyMiddleware, createStore, compose, Store } from "redux";
import { combineEpics, createEpicMiddleware, Epic } from "redux-observable";

import { reducers, State } from "./store";
import { isServer } from "./constants";

function getComposeEnhancers(): any {
  if (!isServer) {
    return (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
  }
  return compose;
}

export interface AppStore {
  store: Store;
  runEffects: (effects$?: Array<Epic>) => void;
}

/**
 * Store configuration
 */
export function configureStore(initialState: State = {} as State): AppStore {
  const epicMiddleware = createEpicMiddleware();
  const composeEnhancers = getComposeEnhancers();
  const middlewares = [epicMiddleware];

  const store: Store = createStore(reducers, initialState, composeEnhancers(applyMiddleware(...middlewares)));

  return {
    store,
    runEffects: (effects$?: Array<Epic>) =>
      effects$
        ? epicMiddleware.run(combineEpics(...effects$))
        : // eslint-disable-next-line no-console
          console.warn("Effects are not running, no epics array was provided")
  };
}
