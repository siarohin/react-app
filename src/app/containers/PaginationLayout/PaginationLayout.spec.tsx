import React from "react";
import { Provider } from "react-redux";
import renderer, { ReactTestRenderer } from "react-test-renderer";
import configureStore from "redux-mock-store";

import { State } from "../../core";
import PaginationLayout from "./PaginationLayout";

const mockStore = configureStore([]);

describe("Containers.PaginationLayout.PaginationLayout: ", () => {
  let store: any;
  let component: ReactTestRenderer;

  beforeEach(() => {
    store = mockStore({
      movieList: {
        totalAmount: 3000,
        limit: 30,
        offset: 0
      }
    } as State);

    component = renderer.create(
      <Provider store={store}>
        <PaginationLayout />
      </Provider>
    );
  });

  it("should render component", () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
});
