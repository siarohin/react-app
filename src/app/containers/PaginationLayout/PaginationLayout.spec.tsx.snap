// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Containers.PaginationLayout.PaginationLayout:  should render component 1`] = `
<nav
  aria-label="pagination navigation"
  className="MuiPagination-root app-pagination"
>
  <ul
    className="MuiPagination-ul"
  >
    <li>
      <button
        aria-current="true"
        aria-label="page 1"
        className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page Mui-selected"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        1
      </button>
    </li>
    <li>
      <button
        aria-label="Go to page 2"
        className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        2
      </button>
    </li>
    <li>
      <button
        aria-label="Go to page 3"
        className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        3
      </button>
    </li>
    <li>
      <button
        aria-label="Go to page 4"
        className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        4
      </button>
    </li>
    <li>
      <button
        aria-label="Go to page 5"
        className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        5
      </button>
    </li>
    <li>
      <div
        className="MuiPaginationItem-root MuiPaginationItem-ellipsis"
      >
        â€¦
      </div>
    </li>
    <li>
      <button
        aria-label="Go to page 100"
        className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        100
      </button>
    </li>
  </ul>
</nav>
`;
